### PR TITLE
f-string formatting

### LINKED ISSUE #35: supporting fstring formatting
It would be great if DateTime could support the new fstring formatting like datetime already does:

```
import datetime, DateTime
for now in (datetime.datetime.now(), DateTime.DateTime()):
    try:
        print(type(now), f'{now = :%-d.%-m.%Y %H:%M}')
    except TypeError as e:
        print(type(now), e)
```

prints --> 

```
<class 'datetime.datetime'> now = 26.5.2022 05:45
<class 'DateTime.DateTime.DateTime'> unsupported format string passed to DateTime.__format__
```

### PR DESCRIPTION
**Pull request by georgpfolz**
Fixes #35 

added \_\_format\_\_ method to DateTime: enables formatting of the date in f-strings and with String format() (like in datetime): 

```
from DateTime import DateTime
d = DateTime()
print(f'my formatted date: {d:%d.%m.%Y}')
print('my formatted date: {d:%d.%m.%Y}'.format(d))
```

### CODE REVIEW

**Comment by dataflake** [comment]
Once the failing tests are fixed I'll take a look

**Comment by d-maurer** [comment]
Georg Pfolz wrote at 2022-6-26 23:36 -0700:
> ...
>
>-- Commit Summary --
>
>  * f-string formatting
>
>-- File Changes --
>
>    M src/DateTime/DateTime.py (8)
>    M src/DateTime/tests/test_datetime.py (6)

Please add an entry in `CHANGES.rst`.

**Comment by georgpfolz** [comment]
> Please add an entry in `CHANGES.rst`.
ok!

As I can see, there is already an empty entry for version 4.5. So I guess I'll put the changes in there.

I'll leave the version number in setup.py unchanged, ok?

(Sorry I don't have a lot of experience with PRs)

**Comment by dataflake** [comment]
Things like version numbers in setup.py are changed only by package managers who make releases etc. Developer pull requests should only contain code changes, tests for these changes and a change log entry so people know what has changed from one version to the next.

**Comment by dataflake** [comment]
FYI: f-strings are only available on Python 3.6 and higher. The `__format__` and `format` methods for strings exists on Python 3.5. So you should skip the test on Python 2.7 and pypy (which is equivalent to Python 2.7) and use the explicit `format` method instead of an f-string in the test.

**Comment by jugmac00** [comment]
@georgpfolz Using an older pip version is no issue - the build succeeded, although annoyingly the following message is printed:

> Error: You are using pip version 20.3.4, however version 22.1.2 is available.

I have not seen this before - imho this should be a warning, not an error.

When I have a look at your builds, I identify the following issues:


a) pypy is broken, see https://github.com/zopefoundation/DateTime/runs/7068251649?check_suite_focus=true 

```
Test-module import failures:
Module: DateTime.tests.test_datetime
  File "/home/runner/work/DateTime/DateTime/src/DateTime/tests/test_datetime.py", line 690
    self.assertEqual(result, f'{dt:%-d.%-m.%Y %H:%M}')
                              ^
SyntaxError: invalid syntax (expected ')')
```

I can't help with that.

b) Coverage dropped...

https://github.com/zopefoundation/DateTime/runs/7068251796?check_suite_focus=true

```
TOTAL                                  1540    148    328     59    87.96%
Coverage failure: total of 87.96 is less than fail-under=88.00
ERROR: InvocationError for command /home/runner/work/DateTime/DateTime/.tox/coverage/bin/coverage report -m --fail-under=88 (exited with code 2)
___________________________________ summary ____________________________________
ERROR:   coverage: commands failed
Error: Process completed with exit code 1.
```

This may mean that you added code which is not tested, maybe only a branch which is not tested.

I would run `tox -e coverage` on the local developer machine, and have a look at the generated html coverage report, especially at the code you added.

**Comment by dataflake** [comment]
> The pip warning is strange, it only happens in the py35 test, and it did not happen in one of the earlier tests. I'll look into this before pushing.

Do not spend time on that warning, it simply doesn't matter.

**Comment by georgpfolz** [comment]
OK, I got rid of the coverage-related error: the last one wasn't even in my code:

```
try:
    __file__
except NameError:  # pragma: no cover
    f = sys.argv[0]
else:
    f = __file__
```

Before I push it to zopefoundation, one last question: Because I used GitHub to do the tests, I have a lot of commits in this branch. Is that ok for merging or should I rebase the last commits into one?

**Comment by georgpfolz** [comment]
> The merges into `master` will by default combine all PR
commits into a single one.

Good to know. Pushed.

Thank you everyone for your guidance and your patience! :)

**Comment by jugmac00** [comment]
@georgpfolz FWIW: The pip "error" is actually a bug in the "python-versions" GitHub action, which has been fixed for Python 3.8 and higher.

https://github.com/actions/python-versions/pull/117

**Comment by dataflake** [comment]
@georgpfolz I have made some small changes to the change log entry so it refers to the specific GitHub issue, and I changed how and when the Python version is checked during unit testing. Now the entire test will be skipped under Python 2.7/pypy, and I re-used an existing definition for Python 3 higher up in the file. No need to re-invent the wheel.

What you normally do when your PR is ready for review, meaning once the unit tests succeed, is to assign reviewers. When I am unfamiliar with who the package developers are I look at the list of recent commits and pick people from there. Once you get approvals from the review you click on the merge button yourself. You will notice it's unavailable right now - the review approvals need to come first.

**Comment by georgpfolz** [comment]
@dataflake 
> What you normally do when your PR is ready for review, meaning once the unit tests succeed, is to assign reviewers. When I am unfamiliar with who the package developers are I look at the list of recent commits and pick people from there. Once you get approvals from the review you click on the merge button yourself. You will notice it's unavailable right now - the review approvals need to come first.

Ok, I assigned some reviewers.

Learned a lot today! :)

**Review by dataflake — APPROVED** [review]

File: src/DateTime/tests/test_datetime.py
Comment (icemac) [inline src/DateTime/tests/test_datetime.py:691]: Comparing with an actual result string makes it easier to read the test and to see that it actually does what it is expected.

Additionally `__format__` has some code branches which are not covered here (`fmt` is no string or an empty string.)  Please write additional tests for them.

File: src/DateTime/tests/test_datetime.py
Comment (icemac) [inline src/DateTime/tests/test_datetime.py:45]: I am not sure if this dance is any more needed at all. But I think that's not part of this PR.

**Review by icemac — CHANGES_REQUESTED** [review]
Thank you for your pull request. Please see the comments I made.

File: src/DateTime/DateTime.py
Comment (georgpfolz) [reply src/DateTime/DateTime.py:1802]: ok!

File: src/DateTime/tests/test_datetime.py
Comment (georgpfolz) [reply src/DateTime/tests/test_datetime.py:691]: > Comparing with an actual result string makes it easier to read the test and to see that it actually does what it is expected.

That's what I did initially but changed it according to a suggestion from dataflake. The problem with the actual f-string is that it produces errors in the tests for earlier Python versions.

I initially also had a check for the Python version, but this created problems with coverage. As far as I understand, I can't fine-tune the pragma comment for coverage, but I can at least allow the test only for Python 3.

**Comment by georgpfolz** [comment]
While doing the requested changes, I thought that after all it would make sense to check the actual f-strings also. My reasoning is that basically the tests using String's format method would also work without the new \_\_format\_\_ method I added to DateTime.

The initial idea of not using the f-strings in the test was not to get an error in the pypy test. 

So I tried to exclude pypy from the f-strings tests: `if not hasattr(sys, 'pypy_version_info')`, but that doesn't seem to work. Obviously, I cannot test this locally, as I don't have pypy installed.

Is there a way to check for pypy, or should I just not test the f-strings?

**Comment by d-maurer** [comment]
Georg Pfolz wrote at 2022-6-28 05:28 -0700:
> ...
>Is there a way to check for pypy, or should I just not test the f-strings?

That `pypy` does not work is due to the fact that it is Python 2.7
(f-strings are only available in Python 3).
You can check for the Python version via a `sys` attribute
or (if `DateTime` already depends in `six`) a standard `six` definition.

**Comment by dataflake** [comment]
The code is not executed. The failure happens earlier when the module is parsed. So it doesn't help trying to check for the Python version, the test module will always fail to import.

**Comment by georgpfolz** [comment]
> The code is not executed. The failure happens earlier when the module is parsed. So it doesn't help trying to check for the Python version, the test module will always fail to import.

Ah, that makes sense!

Thank you, that helped me resolve the problem!

File: src/DateTime/tests/test_datetime.py
Comment (icemac) [inline src/DateTime/tests/test_datetime.py]:
```suggestion
            eval("self.assertEqual(str(dt), f'{dt:}'.format(dt))")
```

File: src/DateTime/tests/test_datetime.py
Comment (icemac) [inline src/DateTime/tests/test_datetime.py]: What's the reason behind using an fstring together with `format`?
I'd expect that the fstring already does the formatting and the `format` call does nothing.

File: src/DateTime/tests/test_datetime.py
Comment (icemac) [inline src/DateTime/tests/test_datetime.py:695]: By default we run the coverage tests on Python 3, so the `no cover` is not needed here. I'd suggest to use the `PY3` mark, so this can be implemented without `eval` when Python 2 support gets dropped.
```suggestion
        if sys.version_info > (3, 6, 0):  # PY3
```

File: src/DateTime/tests/test_datetime.py
Comment (icemac) [inline src/DateTime/tests/test_datetime.py]: This is not well readable, I'd expect something like: (here and for all the other assertions in this test)

```python
self.assertEqual('2022-06-30 08:12:34', '{}'.format(dt))
```

This would require to use a static point in time in line 688 instead of using the current date, but this is fine as we do not test the `now` and `__str__` methods here.

Comment (icemac) [reply src/DateTime/tests/test_datetime.py]: Okay, I see, you implemented it while I was writing. 😀

File: src/DateTime/tests/test_datetime.py
Comment (icemac) [inline src/DateTime/tests/test_datetime.py:698]: What's the reason behind using an fstring together with `format`?
I'd expect that the fstring already does the formatting and the `format` call does nothing.

File: src/DateTime/tests/test_datetime.py
Comment (dataflake) [inline src/DateTime/tests/test_datetime.py:700]:
```suggestion
```

The else clause can be left out entirely.

**Review by icemac — APPROVED** [review]

**Comment by icemac** [comment]
I just released a new version including this PR as https://pypi.org/project/DateTime/4.5/.

**Pull request merged**
